src/http/http-client.js

```diff
@@ -0,0 +1,218 @@
+
+        // only retry 5xx errors and errors that don't have a status code (which
+        // indicates some kind of network or I/O error)
+        if (status && (status < 500 || status > 599)) {
+            throw e;
+        }
+    }
+    if (httpResult) {
+        httpResult.addRetryError(e);
+    }
+}
+
+/**
+ * Builds an object containing retry options that will be provided to the utility method
+ * that submits HTTP requests. The options will include the number of times to retry,
+ * the amount of time between retries, and a method to invoke whenever there is an
+ * intermediate error.
+ *
+ * @param {DirectBinaryUploadOptions} uploadOptions Options controlling the upload process.
+ * @param {HttpResult} httpResult Result being provided for the current operation. Any
+ *   intermediate retry errors will be added to the result.
+ */
+function getRetryOptions(uploadOptions, httpResult) {
+    return {
+        retryCount: uploadOptions.getHttpRetryCount(),
+        retryDelay: uploadOptions.getHttpRetryDelay(),
+        onRetryError: e => handleRetryError(e, httpResult),
+    }
+}
+
+/**
+ * Registers a cancel token, which is used to cancel an HTTP request, with
+ * a corresponding cancel ID.
+ *
+ * @param {string} cancelId ID that will be passed to cancel().
+ * @param {string} tokenId A unique identifier that can be used to identify
+ *  this specific token. Will be used to remove the token when the request
+ *  is finished.
+ * @param {*} cancelToken Axios token that is used to cancel an request.
+ */
+function registerCancelId(cancelId, tokenId, cancelToken) {
+    if (!this.cancelTokens[cancelId]) {
+        this.cancelTokens[cancelId] = [];
+    }
+    this.cancelTokens[cancelId].push(cancelToken);
+}
+
+/**
+ * Removes a cancel token that was previously registered using registerCancelId.
+ *
+ * @param {string} cancelId ID that will be passed to cancel().
+ * @param {string} tokenId A unique identifier that can be used to identify
+ *  the specific token to remove.
+ */
+function unregisterCancelId(cancelId, tokenId) {
+    if (this.cancelTokens[cancelId]) {
+        for (let i = 0; i < this.cancelTokens[cancelId].length; i++) {
+            const { tokenId: currTokenId } = this.cancelTokens[cancelId][i];
+
+            if (tokenId === currTokenId) {
+                this.cancelTokens[cancelId].splice(i, 1);
+                break;
+            }
+        }
+    }
+}
+
+/**
+ * Cancels HTTP requests based on event data from DirectBinaryUploadController's
+ * 'cancel' event data. Depending on event data, the client will either cancel
+ * only those requests whose cancel ID matches the given file name, or cancel
+ * all requests if no fileName is given.
+ *
+ * @param {object} cancelEventData Event data from cancel event.
+ */
+function cancelFromController(cancelEventData) {
+    const { fileName } = cancelEventData;
+
+    if (fileName) {
+        // use fileName as cancel ID and cancel any matches
+        this.cancel(fileName);
```

Comment:
Are filenames unique? (is it a full path?)

Reply:
That's a great question. From the `HttpClient` perspective, I don't think it needs to care. I'll need to think of the implementation's perspective, though. Before these changes we were only uploading files from a single directory, so there wasn't a risk for file name collisions. Now we'll have multiple directories, though, so I'll need to investigate. I do think using the full path would address that issue, so I'll look at making that change.

Reply:
I double-checked this, and made a change to use the file's path instead of its name. I updated the documentation to clarify this. I left the name of the element as `fileName` for backward-compatibility, but renamed the value everywhere I could to clarify that it's the file's path, and not its name.